pkg/registry/rbac/reconciliation/reconcile_clusterrolebindings.go

```diff
@@ -0,0 +1,239 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reconciliation
+
+import (
+	"fmt"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/rbac"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/rbac/internalversion"
+)
+
+type ReconcileClusterRoleBindingOptions struct {
+	RoleBinding *rbac.ClusterRoleBinding
+	Confirm     bool
+	Union       bool
+	Client      internalversion.ClusterRoleBindingInterface
+}
+
+type ReconcileClusterRoleBindingResult struct {
+	// Role is the resulting role binding from the reconciliation operation
+	RoleBinding *rbac.ClusterRoleBinding
+
+	// MissingSubjects contains expected rules that were missing from the currently persisted role
+	MissingSubjects []rbac.Subject
+	// ExtraSubjects contains extra permissions the currently persisted role had
+	ExtraSubjects []rbac.Subject
+
+	// Operation is the API operation required to reconcile
+	Operation ReconcileOperation
+	// Protected indicates an existing role prevented reconciliation
+	Protected bool
+}
+
+func (o *ReconcileClusterRoleBindingOptions) Run() (*ReconcileClusterRoleBindingResult, error) {
+	var result *ReconcileClusterRoleBindingResult
+
+	existingBinding, err := o.Client.Get(o.RoleBinding.Name, metav1.GetOptions{})
+	switch {
+	case errors.IsNotFound(err):
+		result = &ReconcileClusterRoleBindingResult{
+			RoleBinding:     o.RoleBinding,
+			MissingSubjects: o.RoleBinding.Subjects,
+			Operation:       ReconcileCreate,
+		}
+
+	case err != nil:
+		return nil, err
+
+	default:
+		result, err = changedRoleBinding(existingBinding, o.RoleBinding, o.Union)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// If reconcile-protected, short-circuit
+	if result.Protected {
+		return result, nil
+	}
+	// If we're in dry-run mode, short-circuit
+	if !o.Confirm {
+		return result, nil
+	}
+
+	switch result.Operation {
+	case ReconcileRecreate:
+		// Try deleting
+		err := o.Client.Delete(
+			existingBinding.Name,
+			&metav1.DeleteOptions{Preconditions: &metav1.Preconditions{UID: &existingBinding.UID}},
+		)
+		switch {
+		case err == nil:
```

Comment:
I think this would be easier to read as 

```
case err == nil,
errors.IsNotFound(errr),
```

The `fallthrough` looks like a mistake since the options are mutually exclusive and nothing happens until the end.